### PR TITLE
Skip buffer zero initialization for Metal backend

### PR DESCRIPTION
**Pull request by jinleili**
**Connections**
https://github.com/gfx-rs/wgpu/issues/1904

**Testing**
Tested wgpu examples on M1 Mac, Intel Mac and iPhone 12.

### CODE REVIEW

File: wgpu-core/src/command/memory_init.rs [hunk 1]
```diff
@@ -148,11 +148,15 @@ pub(crate) fn fixup_discarded_surfaces<
 
 impl<A: hal::Api> BakedCommands<A> {
     // inserts all buffer initializations that are going to be needed for executing the commands and updates resource init states accordingly
+    #[allow(unused_variables, unreachable_code)]
     pub(crate) fn initialize_buffer_memory(
         &mut self,
         device_tracker: &mut TrackerSet,
         buffer_guard: &mut Storage<Buffer<A>, id::BufferId>,
     ) -> Result<(), DestroyedBufferError> {
+        #[cfg(metal)]
```
Comment (kvark) [inline wgpu-core/src/command/memory_init.rs:157]: Let's implement it properly. wgpu-core doesn't need any platform-specific hacks outside of the backend selection process.
A proper way would be for the wgpu-hal backend to offer a feature "my allocated buffers are always initialized to zero".  And then wgpu-core would check for this feature and activate it.

**Review by kvark — CHANGES_REQUESTED** [review]

File: wgpu-types/src/lib.rs [hunk 2]
```diff
@@ -569,6 +569,12 @@ bitflags::bitflags! {
         ///
         /// This is a native-only feature.
         const TEXTURE_COMPRESSION_ASTC_HDR = 1 << 43;
+        /// Skip buffer zero initialization
+        /// Supported Platforms:
+        /// - Metal
+        ///
+        /// This is a native-only feature.
+        const ALLOCATED_BUFFER_ALREADY_ZERO_FILLED = 1 << 44;
```
Comment (cwfitzgerald) [inline wgpu-types/src/lib.rs:577]: I don't think this should end up in the public facing feature set here as this is something that the user doesn't need to know about at all. @kvark thoughts on where it should go instead?

Comment (kvark) [reply wgpu-types/src/lib.rs:577]: I think it's fine for some of the internal (i.e. needed on the hal - core boundary only) flags to be in this public bitflags list, as long as wgpu-core never reports them back to the user.
We could improve this by moving those internal flags into a separate bitflags type that would live in wgpu-hal.

Comment (jinleili) [reply wgpu-types/src/lib.rs:577]: @kvark Not sure if I understand correctly, is the internal flags something like this?
```rust
bitflags::bitflags! {
	#[repr(transparent)]
	#[derive(Default)]
	pub struct InternalFeatures: u64 {
		const ALLOCATED_BUFFER_ALREADY_ZERO_FILLED = 1 << 0;
	}
}
// wgpu_hal/lib.rs
pub trait Device<A: Api>: Send + Sync {
	...
	fn contains_internal_feature(&self, _feature: wgt::InternalFeatures) -> bool;
}
// wgpu_hal/metal/mod.rs
pub struct Device {
	...
	internal_features: wgt::InternalFeatures,
}

// wgpu_hal/metal/device.rs
impl crate::Device<super::Api> for super::Device {
	...
	fn contains_internal_feature(&self, feature: wgt::InternalFeatures) -> bool {
		self.internal_features.contains(feature)
	}
}
```

Comment (kvark) [reply wgpu-types/src/lib.rs:577]: Let's discuss it as a follow-up, specifically aimed at the flags. Current PR should go in by just using the `Features` and having one there.

**Pull request closed without merging**